src/QsCompiler/Core/SymbolTable.fs

```diff
@@ -980,22 +981,47 @@ and NamespaceManager
                 elif tId |> isBuiltIn BuiltIn.EntryPoint then
                     match box decl.Defined with
                     | :? CallableSignature as signature when not (signature.TypeParameters.Any()) ->
-                        let validateArgAndReturnTypes preventTuples (qsType : QsType) =
+
+                        // verify that the entry point has only a default body specialization
+                        let hasCharacteristics = signature.Characteristics.Characteristics |> function | EmptySet | InvalidSetExpr -> false | _ -> true
+                        match Namespaces.TryGetValue parent.Namespace with
+                        | false, _ -> ArgumentException "no namespace with the given name exists" |> raise
+                        | true, ns when not ((ns.SpecializationsDefinedInAllSources parent.Name).Any(fst >> (<>)QsBody) || hasCharacteristics) -> ()
+                        | _ -> errs.Add (decl.Position, signature.Characteristics.Range.ValueOr decl.Range |> QsCompilerDiagnostic.Error (ErrorCode.InvalidEntryPointSpecialization, []))
+
+                        // validate entry point argument and return type
+                        let validateArgAndReturnTypes argumentRestriction (qsType : QsType) =
```

Comment:
`argumentRestriction` is somewhat misleading; maybe something like `isInArg`.

Alternatively, this could be two functions `validateArgType` and `validateReturnType`, where `validateArgType` calls `validateReturnType` in addition to checking the argument-specific restrictions.

Reply:
renamed